storage/innobase/fsp/fsp0fsp.cc

```diff
@@ -3577,23 +3643,408 @@ dberr_t fsp_sys_tablespace_validate()
+  if (!block)
+    return err;
+  if (!inodes->insert_seg(block->page.frame + PAGE_DATA))
+    err= DB_CORRUPTION;
+
+  block->page.unfix();
+
+  /* Get rollback segment header page */
+  for (ulint rseg_id= 0; rseg_id < TRX_SYS_N_RSEGS && err == DB_SUCCESS;
+       rseg_id++)
+  {
+    trx_rseg_t *rseg= &trx_sys.rseg_array[rseg_id];
+    if (rseg->space->id == 0)
+    {
+      block= buf_pool.page_fix(rseg->page_id(), &err,
+                               buf_pool_t::FIX_WAIT_READ);
+      if (!block)
+        continue;
+      if (!inodes->insert_seg(block->page.frame + TRX_RSEG +
+                              TRX_RSEG_FSEG_HEADER))
+        err= DB_CORRUPTION;
+      block->page.unfix();
+
+      /* Even after slow shutdown, there is a possiblity that
+      cached undo log can exist. So store the segment as used one */
+      for (trx_undo_t *undo= UT_LIST_GET_FIRST(rseg->undo_cached);
+           undo && err == DB_SUCCESS;
+           undo= UT_LIST_GET_NEXT(undo_list, undo))
+      {
+        block= buf_pool.page_fix(page_id_t{0, undo->hdr_page_no}, &err,
+                                 buf_pool_t::FIX_WAIT_READ);
+        if (!block)
+	  continue;
+	if (!inodes->insert_seg(block->page.frame + TRX_UNDO_SEG_HDR +
+                                TRX_UNDO_FSEG_HEADER))
+          err= DB_CORRUPTION;
+        block->page.unfix();
+      }
+    }
+  }
+
+  if (err == DB_SUCCESS)
+    err= fsp_table_inodes(inodes, mtr);
+  return err;
+}
+
+/** Free the extents, fragment page from the given inode
+@param space  system tablespace
+@param page_no index node page number
+@param offset  index node offset within page
+@return error code */
+static dberr_t fseg_inode_free(fil_space_t *space, uint32_t page_no,
+                               uint16_t offset)
+{
+  dberr_t err= DB_SUCCESS;
+  fsp_old_inode old_inodes_entry;
+  mtr_t mtr;
+  mtr.start();
+  mtr.x_lock_space(space);
+  sql_print_information("InnoDB: Freeing the unused segment page "
+                        UINT32PF " offset " UINT32PF, page_no,
+                        offset);
+  old_inodes_entry.insert(0, &mtr);
+  old_inodes_entry.insert(page_no, &mtr);
+
+  uint32_t old_free_len= space->free_len;
+  buf_block_t *iblock= mtr.get_already_latched(
+    page_id_t{0, page_no}, MTR_MEMO_PAGE_SX_FIX);
+  if (!iblock)
+  {
+    err= DB_ERROR;
+err_exit:
+    sql_print_warning("InnoDB: Failed to free the unused segment");
+    old_inodes_entry.restore(&mtr);
+    mtr.commit();
+    space->free_len= old_free_len;
+    return err;
+  }
+
+  fseg_inode_t *inode= iblock->page.frame + offset;
+  xdes_t *descr= fseg_get_first_extent(inode, space, &mtr, &err);
+  while (descr && err == DB_SUCCESS)
+  {
+    old_inodes_entry.insert_descr(descr, &mtr);
+    err= fseg_free_extent(inode, iblock, space,
+                          xdes_get_offset(descr), &mtr);
+    DBUG_EXECUTE_IF("unused_undo_free_fail_4",
+                    err= DB_CORRUPTION;);
+    if (err)
+      goto err_exit;
```

Comment:
Can we split `fseg_free_extent()` to two separate: checking the consistency, and modifying the data? We should first only check the consistency.